doc/config-replace/SONiC_Configuration_Replace_Design_Doc.md

```diff
@@ -0,0 +1,537 @@
+
+
+# 2 Functionality
+## 2.1 Target Deployment Use Cases
+
+The *config replace* features allows switch administrators to push an updated configuration file to a switch with least possible network disruptions. Only the exact configuration change is detected and applied.
+
+During the course of administrating a SONiC switch, the user executes various CLI commands. Some of these commands are targeted to perform a temporary action which needs to be undone later. The *config replace* features allows a user to automatically undo the configurations made without requiring to remember what configuration changes had been performed.
+
+
+
+The *config replace* feature can also illustrate the difference in the running configuration compared to a saved configuration file.
+
+
+
+The existing *config save* command  along with the *config replace* command can be used to perform configuration commit and automatic rollback operations. The user can save the configuration of a SONiC switch at various points in time and can later chose to move to an intended previously known and tested configuration state.
+
+
+
+## 2.2 Functional Description
+The SONiC configuration replace feature can be broadly classified into following sequence of operations:
+
+
+### Stage-1: Patch Creation
+
+Running configuration and user specified config_db.json file are compared and a JSON patch is created.
+
+
+
+### Stage-2: Patch Ordering
+
+As per the JSON standard, the elements described in JSON format are  unordered. However, in SONiC there are some tables which can not be processed before another table is updated. For example changes to the VLAN_MEMBER table needs to be processed before the PORTCHANNEL table is processed. If the ordering is not established, the Port Channel delete operation will fail with an error message complaining that the port-channel is still part of a VLAN. So it is essential that the dependency between the tables is identified and the patch is processed in that order.
+
+
+
+The config replace command uses an ordering table which specifies the list of dependent tables that need to be processed before a table is processed. Below is an example depiction of two ConfigDB tables.
+
+```json
+{
+  "VLAN" : ["VLAN_MEMBER", "VLAN_INTERFACE"],
+  "PORTCHANNEL" : ["VLAN_MEMBERSHIP", "PORTCHANNEL_MEMBER"]
+}
+```
+
+A dependency graph is created for all the tables which are in the JSON patch file. The dependency graph  is then resolved to find the order in which the updates to tables in the JSON patch are performed. The JSON patch file is then sorted in the resolved order. 
+
+
+
+### Stage-3: Patch Apply
+
+Once the patch apply order has been determined, the patch file is applied. 
+
+
+
+The ConfigDB JSON patch file is a list of elements which contain following fields.  Each element results in one or more ConfigDB updates.  For detailed information about the JSON patch operations refer to the  section 4(Operations) of [RFC 6902](https://tools.ietf.org/html/rfc6902).
```

Comment:
Can it possible to define a config replace contract so that all the features must implement it wihtout miss it. The contract could be the steps pre/depency graph/config-replace/post/corrective action/enable or disable.